sched/semaphore/sem_trywait.c

```diff
@@ -72,7 +72,7 @@ int nxsem_trywait(FAR sem_t *sem)
 
   /* This API should not be called from interrupt handlers & idleloop */
 
-  DEBUGASSERT(sem != NULL && up_interrupt_context() == false);
```

Comment:
Relaxing overly strict assertion. trywait does not block so it's safe to be called from interrupt handlers. https://www.mail-archive.com/commits@nuttx.apache.org/msg44549.html

Reply:
Why would interrupts need semaphore locking ? Semaphores are a user space API.

More specifically, what should the ISR procedure that needs a semaphore lock do, in case the semaphore is locked ? I assume it was going to do something but was prevented from doing it ?

Reply:
same same. I think this debugassert is valid, it is a bug to call this from isr. Where does this bite?

Reply:
Hi, it's needed in an ISR where a frame is needed to be sent to the socketCAN interface so that higher level user space application can receive the RX packet (via callback). In this `can_input` func, it needs to call `can_callback` to notify the user application listener about the RX packet. There it needs to lock, send event and unlock the network.
https://github.com/tiiuae/nuttx/blob/c64610f711eb7881808f7d71ea5a1f64d5807137/net/can/can_input.c#L176
```
      /* Send to socket interface */

      NETDEV_RXPACKETS(&priv->dev);
      can_input(&priv->dev);
```

Reply:
ofc, this `can_input` for sending can frame to user application can be delegated to a seperate work queue but for high performance, ideally the RX frame needs to be forwarded upstream to user application asap there is ISR triggered for RX.
And it's programmed that way in all other SocketCAN compatible CAN driver in Nuttx
For ex:
https://github.com/tiiuae/nuttx/blob/c64610f711eb7881808f7d71ea5a1f64d5807137/arch/arm/src/stm32h7/stm32_fdcan_sock.c#L1197

Reply:
Hi, this looks like awful lots of code to be called from an ISR anyway. Maybe it would be better to schedule a work queue item from the ISR, from which you can then call the can_input ?

Reply:
please try to schedule a work queue item into high priority wq. I don't think it can even work properly when called directly from the ISR, it tries to allocate IOBs in some branches of the code later one.

Reply:
anyway, I can try putting the receiving functionality to a separate work queue function and test if it would compromise the performance